src/host/multiple_ledger.h

```diff
@@ -0,0 +1,805 @@
+
+      // Header reserved for the offset to the position table
+      fseeko(file, sizeof(positions_offset_header_t), SEEK_SET);
+      total_len = sizeof(positions_offset_header_t);
+    }
+
+    // Used when recovering an existing ledger file
+    LedgerFile(const std::string& dir, const std::string& file_name) :
+      dir(dir),
+      file(NULL)
+    {
+      auto full_path = (fs::path(dir) / fs::path(file_name));
+      file = fopen(full_path.c_str(), "r+b");
+      if (!file)
+      {
+        throw std::logic_error(
+          fmt::format("Unable to open ledger file {}", full_path));
+      }
+
+      compacted = is_ledger_file_compacted(file_name);
+      start_idx = get_start_idx_from_file_name(file_name);
+
+      // First, get full size of file
+      fseeko(file, 0, SEEK_END);
+      size_t total_file_size = ftello(file);
+
+      // Second, read offset to header table
+      fseeko(file, 0, SEEK_SET);
+      positions_offset_header_t table_offset;
+      if (fread(&table_offset, sizeof(positions_offset_header_t), 1, file) != 1)
+      {
+        throw std::logic_error(fmt::format(
+          "Failed to read positions offset from ledger file {}", full_path));
+      }
+
+      if (table_offset != 0)
+      {
+        // If the chunk was completed, read positions table from file directly
+        total_len = table_offset;
+        fseeko(file, table_offset, SEEK_SET);
+
+        positions.resize(
+          (total_file_size - table_offset) / sizeof(positions.at(0)));
+
+        if (
+          fread(
+            positions.data(),
+            sizeof(positions.at(0)),
+            positions.size(),
+            file) != positions.size())
+        {
+          throw std::logic_error(fmt::format(
+            "Failed to read positions table from ledger file {}", full_path));
+        }
+        completed = true;
+      }
+      else
+      {
+        // If the chunk was not completed, read all entries to reconstruct
+        // positions table
+        total_len = total_file_size;
+
+        auto len = total_len - sizeof(positions_offset_header_t);
+        size_t pos = sizeof(positions_offset_header_t);
+        uint32_t entry_size = 0;
+
+        while (len >= frame_header_size)
+        {
+          if (fread(&entry_size, frame_header_size, 1, file) != 1)
+          {
+            throw std::logic_error(fmt::format(
+              "Failed to read frame from ledger file {}", full_path));
+          }
+
+          len -= frame_header_size;
+
+          if (len < entry_size)
+          {
+            throw std::logic_error(
+              fmt::format("Malformed ledger file {}", full_path));
+          }
+
+          fseeko(file, entry_size, SEEK_CUR);
+          len -= entry_size;
+
+          positions.push_back(pos);
+          pos += (entry_size + frame_header_size);
+        }
+        completed = false;
+      }
+    }
+
+    ~LedgerFile()
+    {
+      fflush(file);
+      fclose(file);
```

Comment:
```suggestion
      fclose(file);
```

Explicit flushing is unnecessary, fclose already does that: https://linux.die.net/man/3/fclose